Tests/otlLib/builder_test.py

```diff
@@ -1392,6 +1392,63 @@ def test_stat_infinities():
     assert struct.pack(">l", posInf) == b"\x7f\xff\xff\xff"
 
 
+class ChainContextualRulesetTest(object):
+    def test_makeRulesets(self):
+        font = ttLib.TTFont()
+        font.setGlyphOrder(["a","b","c","d","A","B","C","D","E"])
+        sb = builder.ChainContextSubstBuilder(font, None)
+        prefix = [ ["a"], ["b"] ]
+        input_ = [ ["c"] ]
+        suffix = [ ]
+        lookups = [ None ]
+        sb.rules.append((prefix, input_, suffix, lookups))
+
+        prefix = [ ["a"], ["d"] ]
+        sb.rules.append((prefix, input_, suffix, lookups))
+
+        sb.add_subtable_break(None)
+
+        # Second subtable has some glyph classes
+        prefix = [ ["A"] ]
+        input_ = [ ["E"] ]
+        sb.rules.append((prefix, input_, suffix, lookups))
+        input_ = [ ["C","D"] ]
+        sb.rules.append((prefix, input_, suffix, lookups))
+        prefix = [ ["A", "B"] ]
+        input_ = [ ["E"] ]
+        sb.rules.append((prefix, input_, suffix, lookups))
+
+        sb.add_subtable_break(None)
+
+        # Third subtable has no pre/post context
+        prefix = []
+        suffix = []
+        sb.rules.append((prefix, input_, suffix, lookups))
+        input_ = [ ["C","D"] ]
+        sb.rules.append((prefix, input_, suffix, lookups))
```

Comment:
it's hard to track the state of the prefix, input_, suffix local variables as you mutate them inside the test. maybe spell out the content of each every time you append a new rule (even if it means repeating the same string literals) so it's clear what rules contain what.